File: sdks/typescript/src/core/pq-checkpoint.ts
/**
 * DCP v2.0 Lazy PQ Checkpoint model.
 *
 * Each audit event is signed with Ed25519 only (microsecond-scale). Every N
 * events, a PQ checkpoint is produced: a composite signature over the Merkle
 * root of the batch. This gives real-time classical security with periodic
 * PQ assurance, reducing PQ signature count by the checkpoint interval factor.
 */

import { randomUUID } from 'crypto';
import type { PQCheckpoint, AuditEventV2 } from '../types/v2.js';
import type { CompositeKeyPair } from './composite-ops.js';
import { compositeSign } from './composite-ops.js';
import { AlgorithmRegistry } from './crypto-registry.js';
import { DCP_CONTEXTS } from './domain-separation.js';
import { sha256Hex } from './dual-hash.js';
import { canonicalizeV2 } from './canonicalize.js';
import type { SecurityTier } from './security-tier.js';
import { tierToCheckpointInterval } from './security-tier.js';

/**
 * Compute a Merkle root from audit event payloads (SHA-256).
 * Each leaf is sha256(canonicalize(event)).
 */
export function auditEventsMerkleRoot(events: AuditEventV2[]): string {
  if (events.length === 0) {
    throw new Error('Cannot compute Merkle root of empty event list');
  }

  let leaves = events.map((e) => {
    const canonical = canonicalizeV2(e);
    return sha256Hex(Buffer.from(canonical, 'utf8'));
  });

  while (leaves.length > 1) {
    if (leaves.length % 2 === 1) leaves.push(leaves[leaves.length - 1]);
    const next: string[] = [];
    for (let i = 0; i < leaves.length; i += 2) {
      const combined = Buffer.concat([
        Buffer.from(leaves[i], 'hex'),
        Buffer.from(leaves[i + 1], 'hex'),
      ]);
      next.push(sha256Hex(combined));
    }
    leaves = next;
  }

  return leaves[0];
}

/**
 * Create a PQ checkpoint for a batch of audit events.
 *
 * @param registry - Algorithm registry with registered providers
 * @param events - The batch of audit events to checkpoint
 * @param sessionNonce - The session nonce shared by all events
 * @param keys - Classical + PQ keypair for composite signing
 * @returns A signed PQCheckpoint
 */
export async function createPQCheckpoint(
  registry: AlgorithmRegistry,
  events: AuditEventV2[],
  sessionNonce: string,
  keys: CompositeKeyPair,
): Promise<PQCheckpoint> {
  if (events.length === 0) {
    throw new Error('Cannot create checkpoint for empty event list');
  }

  const merkleRoot = auditEventsMerkleRoot(events);
  const checkpointId = `ckpt-${randomUUID()}`;

  const checkpointPayload = {
    checkpoint_id: checkpointId,
    session_nonce: sessionNonce,
    event_range: {
      from_audit_id: events[0].audit_id,
      to_audit_id: events[events.length - 1].audit_id,
      count: events.length,
    },
    merkle_root: `sha256:${merkleRoot}`,
  };

  const canonical = canonicalizeV2(checkpointPayload);
  const payloadBytes = new TextEncoder().encode(canonical);

  const compositeSig = await compositeSign(
    registry,
    DCP_CONTEXTS.AuditEvent,
    payloadBytes,
    keys,
  );

  return {
    ...checkpointPayload,
    composite_sig: compositeSig,
  };
}

/**
 * Manager for tracking audit events and producing PQ checkpoints at
 * configurable intervals.
 *
 * The interval can be set explicitly or derived from a {@link SecurityTier}.
 * When a tier is provided without an explicit interval, the interval is
 * computed via {@link tierToCheckpointInterval}.
 */
export class PQCheckpointManager {
  private pendingEvents: AuditEventV2[] = [];
  private checkpoints: PQCheckpoint[] = [];
  private _interval: number;
  private _tier: SecurityTier | undefined;

  constructor(
    interval: number,
    private readonly registry: AlgorithmRegistry,
    private readonly sessionNonce: string,
    private readonly keys: CompositeKeyPair,
    tier?: SecurityTier,
  ) {
    if (tier !== undefined) {
      this._tier = tier;
      this._interval = tierToCheckpointInterval(tier);
    } else {
      this._interval = interval;
    }
    if (this._interval < 1) throw new Error('Checkpoint interval must be >= 1');
  }

  /** Current checkpoint interval. */
  get interval(): number {
    return this._interval;
  }

  /** Current security tier (undefined if constructed with explicit interval). */
  get tier(): SecurityTier | undefined {
    return this._tier;
  }

  /**
   * Update the security tier at runtime, recalculating the checkpoint interval.
   * Pending events are not flushed — call {@link flush} first if needed.
   */
  setTier(tier: SecurityTier): void {
    this._tier = tier;
    this._interval = tierToCheckpointInterval(tier);
  }

  /**
   * Record an audit event. If the pending count reaches the checkpoint
   * interval, a PQ checkpoint is automatically produced and returned.
   */
  async recordEvent(event: AuditEventV2): Promise<PQCheckpoint | null> {
    this.pendingEvents.push(event);
    if (this.pendingEvents.length >= this._interval) {
      return this.flush();
    }
    return null;
  }

  /**
   * Force a PQ checkpoint over all pending events (e.g. at session end).
   * Returns null if no events are pending.
   */
  async flush(): Promise<PQCheckpoint | null> {
    if (this.pendingEvents.length === 0) return null;

    const checkpoint = await createPQCheckpoint(
      this.registry,
      this.pendingEvents,
      this.sessionNonce,
      this.keys,
    );

    this.checkpoints.push(checkpoint);
    this.pendingEvents = [];
    return checkpoint;
  }

  /** Get all checkpoints produced so far. */
  getCheckpoints(): PQCheckpoint[] {
    return [...this.checkpoints];
  }

  /** Get the count of pending (un-checkpointed) events. */
  getPendingCount(): number {
    return this.pendingEvents.length;
  }
}

File: sdks/python/dcp_ai/v2/pq_checkpoint.py
"""
DCP v2.0 Lazy PQ Checkpoint — Python port.

Each audit event is signed classically (Ed25519). Every N events a PQ
checkpoint is produced: a composite signature over the Merkle root of
the batch. Gives real-time classical security with periodic PQ assurance.
"""

from __future__ import annotations

import hashlib
import uuid
from typing import Any

from dcp_ai.v2.canonicalize import canonicalize_v2
from dcp_ai.v2.composite_ops import CompositeKeyInfo, composite_sign
from dcp_ai.v2.crypto_registry import AlgorithmRegistry
from dcp_ai.v2.domain_separation import DCP_CONTEXTS
from dcp_ai.v2.security_tier import tier_to_checkpoint_interval


def audit_events_merkle_root(events: list[dict[str, Any]]) -> str:
    """Compute a SHA-256 Merkle root over canonicalised event payloads.

    Returns a 64-char hex string (no "sha256:" prefix).
    """
    if not events:
        raise ValueError("Cannot compute Merkle root of empty event list")

    leaves = [
        hashlib.sha256(canonicalize_v2(e).encode("utf-8")).hexdigest() for e in events
    ]
    while len(leaves) > 1:
        if len(leaves) % 2 == 1:
            leaves.append(leaves[-1])
        next_leaves: list[str] = []
        for i in range(0, len(leaves), 2):
            combined = bytes.fromhex(leaves[i]) + bytes.fromhex(leaves[i + 1])
            next_leaves.append(hashlib.sha256(combined).hexdigest())
        leaves = next_leaves
    return leaves[0]


async def create_pq_checkpoint(
    registry: AlgorithmRegistry,
    events: list[dict[str, Any]],
    session_nonce: str,
    classical_key: CompositeKeyInfo,
    pq_key: CompositeKeyInfo,
) -> dict[str, Any]:
    """Create a PQ checkpoint over a batch of audit events."""
    if not events:
        raise ValueError("Cannot create checkpoint for empty event list")

    merkle = audit_events_merkle_root(events)
    checkpoint_id = f"ckpt-{uuid.uuid4()}"

    payload: dict[str, Any] = {
        "checkpoint_id": checkpoint_id,
        "session_nonce": session_nonce,
        "event_range": {
            "from_audit_id": events[0]["audit_id"],
            "to_audit_id": events[-1]["audit_id"],
            "count": len(events),
        },
        "merkle_root": f"sha256:{merkle}",
    }
    canonical = canonicalize_v2(payload)
    composite = await composite_sign(
        registry, DCP_CONTEXTS["AuditEvent"], canonical.encode("utf-8"), classical_key, pq_key
    )
    return {**payload, "composite_sig": composite}


class PQCheckpointManager:
    """Collect audit events and produce PQ checkpoints at a configurable interval."""

    def __init__(
        self,
        interval: int,
        registry: AlgorithmRegistry,
        session_nonce: str,
        classical_key: CompositeKeyInfo,
        pq_key: CompositeKeyInfo,
        tier: str | None = None,
    ) -> None:
        if tier is not None:
            self._tier = tier
            self._interval = tier_to_checkpoint_interval(tier)
        else:
            self._tier = None
            self._interval = interval
        if self._interval < 1:
            raise ValueError("Checkpoint interval must be >= 1")
        self._registry = registry
        self._session_nonce = session_nonce
        self._classical_key = classical_key
        self._pq_key = pq_key
        self._pending: list[dict[str, Any]] = []
        self._checkpoints: list[dict[str, Any]] = []

    @property
    def interval(self) -> int:
        return self._interval

    @property
    def tier(self) -> str | None:
        return self._tier

    def set_tier(self, tier: str) -> None:
        """Update the tier at runtime; does not flush pending events."""
        self._tier = tier
        self._interval = tier_to_checkpoint_interval(tier)

    async def record_event(self, event: dict[str, Any]) -> dict[str, Any] | None:
        """Record an event; when pending reaches interval, emit a checkpoint."""
        self._pending.append(event)
        if len(self._pending) >= self._interval:
            return await self.flush()
        return None

    async def flush(self) -> dict[str, Any] | None:
        """Force a checkpoint over all pending events (e.g. at session end)."""
        if not self._pending:
            return None
        checkpoint = await create_pq_checkpoint(
            self._registry,
            self._pending,
            self._session_nonce,
            self._classical_key,
            self._pq_key,
        )
        self._checkpoints.append(checkpoint)
        self._pending = []
        return checkpoint

    def get_checkpoints(self) -> list[dict[str, Any]]:
        return list(self._checkpoints)

    def get_pending_count(self) -> int:
        return len(self._pending)
